### PR TITLE
Final capstone project: Book an appointment

### PR DESCRIPTION
**Pull request by zkr024**
### CARS APP
An app that allows a user to select a car test drive

[Front-end App](https://github.com/NeckerFree/Cars/tree/dev)

### **Project requirements**

**Basics**

- You should follow the layout of [the provided design](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign). 
- Select a theme for your website 

**Features**

- Core features - for teams of each size

  - The user logs in to the website, only by typing the username 

- In the navigation panel, the user can see links to:

  - Cars that you selected as a theme.

  - "Reserve" form.
  - "My reservations".
  - "Add cars that you selected as a theme" (if you are in a group of 4 people the link is visible to everybody).
  - "Delete cars that you selected as a theme" (if you are in a group of 4 people the link is visible to everybody).
  - On the main page, the user can see a list of cars 
 (https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005577).
 

- List
  - When the user selects a specific item, they can see the [details page](https://www.behance.net/gallery/26425031/Vespa-Responsive-Redesign/modules/173005579) with its full description (skip the "Rotate image" button).
  - On the details page, the user can click the "Reserve" button (in the design you can see the "Configure" button - please replace it with the "Reserve" button).

- Details

  - When the user clicks the "Add Item" link in the navigation panel they can see a form for adding a new item.
  - Make the app responsive, creating both mobile and desktop versions.

**Responsive app preview**

page

**Rspec Testing**
Screenshot from 2022-09-02 02-17-38

### CODE REVIEW

File: config/database.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+#       Install PostgreSQL and put its /bin directory on your path.
+#
+# Configure Using Gemfile
+# gem "pg"
+#
+default: &default
+  adapter: postgresql
+  encoding: unicode
+  username: amrendra
+  password: amrendra
```
Comment (vichuge) [inline config/database.yml]: - Username and password are not required to have a good DB configuration. If you don't declare this, the local installation takes the username/password created locally.

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  resources :users do
+    resources :appointments
+    resources :selections
+    resources :cars
+  end  
+
+  resources :countries do
+    resources :cities
+  end
+  resources :sellers
```
Comment (vichuge) [inline config/routes.rb]: - Not all times are required to use all endpoints available on the resources command. In this case, try using `only:` to open all the required endpoints for each controller.

Example:

`resources :cars only: %i[index show delete]`

File: app/models/appointment.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Appointment < ApplicationRecord
+  belongs_to :user
+  belongs_to :seller
+  belongs_to :car
+  belongs_to :city
```
Comment (vichuge) [inline app/models/appointment.rb:5]: - Please, add validations for elements on this table. I understand not all elements are required but, I kindly invite you to re-review this part and be sure that your validations are enough for this.

File: app/models/car.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class Car < ApplicationRecord
+  has_many :appointments
+
+  has_many :selections, dependent: :destroy
+  has_many :users, through: :selections, dependent: :destroy
```
Comment (vichuge) [inline app/models/car.rb:5]: - Please, add validations for elements on this table. I understand not all elements are required but, I kindly invite you to re-review this part and be sure that your validations are enough for this.

File: app/models/city.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class City < ApplicationRecord
+  belongs_to :country
+
+  has_many :appointments
```
Comment (vichuge) [inline app/models/city.rb:4]: - Please, add validations for elements on this table. I understand not all elements are required but, I kindly invite you to re-review this part and be sure that your validations are enough for this.

File: app/models/country.rb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+class Country < ApplicationRecord
+  has_many :cities
```
Comment (vichuge) [inline app/models/country.rb:2]: - Please, add validations for elements on this table. I understand not all elements are required but, I kindly invite you to re-review this part and be sure that your validations are enough for this.

File: app/models/selection.rb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+class Selection < ApplicationRecord
+  belongs_to :user
+  belongs_to :car
```
Comment (vichuge) [inline app/models/selection.rb:3]: - Please, add validations for elements on this table. I understand not all elements are required but, I kindly invite you to re-review this part and be sure that your validations are enough for this.

File: app/models/seller.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class Seller < ApplicationRecord
+  has_many :appointments
+
+  has_many :users, through: :appointments
```
Comment (vichuge) [inline app/models/seller.rb:4]: - Please, add validations for elements on this table. I understand not all elements are required but, I kindly invite you to re-review this part and be sure that your validations are enough for this.

**Review by vichuge — COMMENTED** [review]
Hi dear student, 

Good job so far!
You still need to work on some issues to prepare your project for the final evaluation, but you are almost there!

mario_ladders


## Suggested changes
_Check the comments under the review._

- Please, consider reviewing the security flows on some endpoints. The users' list at this point is available to all unregistered users. I kindly invite you to fix this.
image

### Please check the [Front-end PR](https://github.com/NeckerFree/Cars/pull/17) to see the other review part. There you can see the improvement for that part of this project.

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Review by mcrd25 — APPROVED** [review]
Hi Team,

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**